tests/phpunit/CRM/Event/Form/ParticipantTest.php

```diff
@@ -268,14 +268,9 @@ public function testSubmitWithFailedPayment(string $thousandSeparator): void {
     $this->setCurrencySeparators($thousandSeparator);
     $paymentProcessorID = $this->processorCreate(['is_test' => 0]);
     $_REQUEST['mode'] = 'live';
-    Civi\Payment\System::singleton()->getById($paymentProcessorID)->setDoDirectPaymentResult(['payment_status_id' => 'failed']);
-    try {
-      $this->submitForm(['is_monetary' => 1, 'financial_type_id' => 1], $this->getSubmitParamsForCreditCardPayment($paymentProcessorID), TRUE);
-    }
-    catch (CRM_Core_Exception_PrematureExitException $e) {
-      return;
-    }
-    $this->fail('should have hit premature exit');
+    \Civi\Payment\System::singleton()->getById($paymentProcessorID)->setDoDirectPaymentResult(['payment_status_id' => 'failed']);
```

Comment:
@eileenmcnaughton The return value stands out here. I *thought* we were only supporting "completed" or "pending" as return value. Also payment_status_id should actually be an ID, payment_status should contain the text. See https://docs.civicrm.org/dev/en/latest/extensions/payment-processors/paymentclass/#dopayment-function

I realise you're not actually changing it here. But it just stood out as wrong. A failed payment should throw an exception

Reply:
@mattwire it is throwing an exception - but that formWrapper is catching it - but we still check for it - see 2 lines further down `$this->assertPrematureExit();` - that is checking the premature exit is thrown as before

Since the form wrapper is tracking mail, variables etc it makes it easier on both sides if it catches the exception too